net/speedtest/speedtest_client.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"net"
+	"time"
+)
+
+// StartClient dials the given address and starts a speedtest.
+// It returns any errors that come up in the tests.
+// It returns an error if the given test type isn't either download or upload.
+// If there are no errors in the test, it returns a slice of results.
+func StartClient(config TestConfig, host, port string) ([]Result, error) {
+	serverAddr, err := net.ResolveTCPAddr("tcp", host+":"+port)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := net.DialTCP("tcp", nil, serverAddr)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	encoder := json.NewEncoder(conn)
+
+	conn.SetReadBuffer(LenBufData)
+	err = encoder.Encode(config)
+	if err != nil {
+		return nil, err
+	}
+	var response TestConfigResponse
+	decoder := json.NewDecoder(conn)
+	err = decoder.Decode(&response)
+	if err != nil {
+		return nil, err
+	}
+	if response.Errors != "" {
+		return nil, errors.New(response.Errors)
+	}
+	return startTestC(conn, config)
+}
+
+// startTestC handles the entire download speed test.
+// It has a loop that breaks if the connection recieves an IO error or if the server sends a header
+// with the "end" type. It reads the headers and data coming from the server and records the number of bytes recieved in each interval in a result slice.
+func startTestC(conn *net.TCPConn, config TestConfig) ([]Result, error) {
+	bufferData := make([]byte, LenBufData)
+	var downloadBegin time.Time
+
+	sum := 0
+	totalSum := 0
+	var lastCalculated float64 = 0.0
+	breakLoop := false
+	var results []Result
+	firstTime := true
+
+	for {
+		n, err := io.ReadFull(conn, bufferData)
+		if firstTime {
```

Comment:
Since it is only used to track downloadBegin initialization, firstTime can be removed and use:
```
if downloadBegin.IsZero() {
    downloadBegin = time.Now()
}
```